### PR TITLE
[risk=low][RW-7973] Async Duplicate Workspace

### PR DESCRIPTION
**Pull request by jmthibault79**
Tested locally.  Duplication took just over a minute - which would have failed in synchronous mode.

Followup to #6375 #6387 #6389

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/main/e2e/README.md#examples)
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: api/src/main/resources/workbench-api.yaml [hunk 1]
```diff
@@ -4783,6 +4826,9 @@ definitions:
         description: The newly created workspace duplicate.
   WorkspaceOperation:
     type: object
+    required:
```
Comment (jmthibault79) [inline api/src/main/resources/workbench-api.yaml:4829]: misc cleanup

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java [hunk 2]
```diff
@@ -237,18 +266,16 @@ private FirecloudWorkspaceId generateFirecloudWorkspaceId(String namespace, Stri
         .orElse(ResponseEntity.notFound().build());
   }
 
-  @Override
-  public ResponseEntity<Void> processCreateWorkspaceTask(CreateWorkspaceTaskRequest request) {
+  private void processWorkspaceTask(long operationId, Supplier<Workspace> workspaceAction) {
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java:269]: generic version of the previous

File: ui/src/app/pages/workspace/workspace-edit.tsx [hunk 3]
```diff
@@ -1055,27 +1056,70 @@ export const WorkspaceEdit = fp.flow(
       this.saveWorkspace();
     }
 
-    private async apiCreateWorkspaceAsync() {
+    private async pollForAsyncWorkspaceOperation(
```
Comment (jmthibault79) [inline ui/src/app/pages/workspace/workspace-edit.tsx:1059]: also a generic version of the previous

File: ui/src/app/pages/workspace/workspace-edit.tsx [hunk 4]
```diff
@@ -1055,27 +1056,70 @@ export const WorkspaceEdit = fp.flow(
...
+            {
+              includeUserRoles: this.state.cloneUserRole,
+              workspace: this.state.workspace,
+            }
+          ),
+        'Workspace duplication failed'
+      );
+    }
+
+    // the API endpoint is synchronous in that it waits for the Terra call
```
Comment (jmthibault79) [inline ui/src/app/pages/workspace/workspace-edit.tsx:1109]: LMK if this comment is just making things worse :)

**Comment by jmthibault79** [comment]
Back to Draft while I fix some unit tests

File: ui/src/app/pages/workspace/workspace-edit.spec.tsx [hunk 5]
```diff
@@ -409,6 +411,12 @@ describe('WorkspaceEdit', () => {
     expect(mockNavigate).toHaveBeenCalledTimes(1);
   });
 
+  // TODO RW-7973
```
Comment (jmthibault79) [inline ui/src/app/pages/workspace/workspace-edit.spec.tsx:414]: add this to the extensive list of TODOs for this ticket - after merge

File: ui/src/testing/stubs/workspaces-api-stub.ts [hunk 6]
```diff
@@ -105,31 +131,67 @@ export class WorkspacesApiStub extends WorkspacesApi {
     });
   }
 
+  private duplicateWorkspaceImpl(
```
Comment (jmthibault79) [inline ui/src/testing/stubs/workspaces-api-stub.ts:134]: extracted from the `cloneWorkspace()` stub so it can also be used by duplicateWorkspaceAsync()

**Comment by jmthibault79** [comment]
e2e test failures are both Duplicate Workspace - which makes sense because they depend on endpoints which don't exist yet on Test (sigh)

**Review by evrii — APPROVED** [review]

**Pull request merged**